geode-core/src/main/java/org/apache/geode/internal/HeapDataOutputStream.java

```diff
@@ -159,32 +161,32 @@ public int sendTo(SocketChannel chan) throws IOException {
     return result;
   }
 
-  public void sendTo(SocketChannel chan, ByteBuffer out) throws IOException {
+  public void sendTo(SocketChannel chan, ByteBuffer out, NioFilter ioFilter)
+      throws IOException {
     finishWriting();
     if (size() == 0) {
       return;
     }
     out.clear();
     if (this.chunks != null) {
       for (ByteBuffer bb : this.chunks) {
-        sendChunkTo(bb, chan, out);
+        sendChunkTo(bb, chan, out, ioFilter);
       }
     }
-    sendChunkTo(this.buffer, chan, out);
-    flushBuffer(chan, out);
+    sendChunkTo(this.buffer, chan, out, ioFilter);
+    flushBuffer(chan, out, ioFilter);
   }
 
   /**
    * sends the data from "in" by writing it to "sc" through "out" (out is used to chunk to data and
    * is probably a direct memory buffer).
    */
-  private void sendChunkTo(ByteBuffer in, SocketChannel sc, ByteBuffer out) throws IOException {
+  private void sendChunkTo(ByteBuffer in, SocketChannel sc, ByteBuffer out, NioFilter ioFilter)
+      throws IOException {
     int bytesSent = in.remaining();
     if (in.isDirect()) {
-      flushBuffer(sc, out);
-      while (in.remaining() > 0) {
-        sc.write(in);
-      }
+      flushBuffer(sc, out, ioFilter);
+      writeToSC(sc, in, ioFilter);
```

Comment:
I am not expert in this area. But from the code changes it looks like we are calling writeToSC twice now (in flushBuffer)...Is this expected?